### PR TITLE
fix: ensure cross-app through dependencies (swev-id: django__django-15973)

### LINKED ISSUE #366: ManyToMany custom through across apps causes AttributeError during migration (swev-id: django__django-15973)
## User Request
Defining the "through" model in a many-to-many field in another app causes "AttributeError: 'str' object has no attribute '_meta'" on migration.

Repository: agyn-sandbox/django
Active branch: django__django-15973

## Observed Failure
- Error: `AttributeError: 'str' object has no attribute '_meta'` during `migrate` when accessing `field.remote_field.through._meta.auto_created` in `django/db/backends/base/schema.py`.
- Moving the `FonteVariavelModel` (the custom through) into the same app as the `ManyToManyField` avoids the error; co-locating all models also avoids the error.

Stack trace (excerpt):
```
Applying fonte.0001_initial...Traceback (most recent call last):
  ...
  File "/.../django/db/backends/base/schema.py", line 453, in create_model
    if field.remote_field.through._meta.auto_created:
AttributeError: 'str' object has no attribute '_meta'
```

## Reproduction Steps
- Create three apps: `fonte`, `variavel`, `fonte_variavel`.
- Models:
  - `fonte.models.FonteModel` with `variaveis = ManyToManyField("variavel.VariavelModel", through="fonte_variavel.FonteVariavelModel")`.
  - `variavel.models.VariavelModel` (simple model).
  - `fonte_variavel.models.FonteVariavelModel` with `ForeignKey('variavel.VariavelModel')` and `ForeignKey('fonte.FonteModel')`.
- Run `makemigrations`. The migration for `fonte` depends only on `('variavel', '__first__')`; it does not include a dependency on `fonte_variavel`.
- Run `migrate`. Error occurs when the schema editor attempts to access `_meta` on `field.remote_field.through` which is still a string because `fonte_variavel` was not migrated first.

## Researcher Specification (by Emerson Gray)
- Root cause: The migrations autodetector incorrectly adds dependencies for a `ManyToManyField` with a custom `through`. It adds dependency to the remote target model but misses the dependency on the `through` model (especially when referenced as a string in another app), leaving `through` unresolved at migration execution time.

- Proposed fix:
  1) Correct dependency resolution in `django/db/migrations/autodetector.py`:
     - In `_get_dependencies_for_foreign_key()`, if `field.remote_field.through` is set, resolve the dependency using the `through` reference (string or class), not the remote target model.
     - Pseudocode change:
       ```python
       if getattr(field.remote_field, "through", None):
           through_app_label, through_object_name = resolve_relation(
               field.remote_field.through, app_label, model_name
           )
           dependencies.append((through_app_label, through_object_name, None, True))
       ```

  2) Optional defensive guard in `django/db/backends/base/schema.py`:
     - Before accessing `through._meta`, resolve string `through` references via `model._meta.apps.get_model(app_label, model_name)`.
     - This prevents `AttributeError` in rare ordering edge cases, though the primary fix should address dependency ordering.

- Tests to add (under `tests/migrations/test_autodetector.py`):
  - Cross-app custom through dependency: ensure `makemigrations` produces a migration for the app with the `ManyToManyField` that depends on the app providing the custom `through`.
  - Ensure dependency points to the `through` app, not only to the remote target model's app.
  - Optional: Guard behavior confirming no crash when `through` is unresolved (if we add the schema editor guard).

- Risks:
  - Adding the correct through dependency may surface new dependency cycles; Django’s migration graph handling should resolve these appropriately.

## Next Steps
- Implement the autodetector fix and add tests in a new branch, then open a PR targeting `django__django-15973` with title including `swev-id: django__django-15973`. We will not run CI per request.

### PR DESCRIPTION
**Pull request by casey-brooks**
## Summary
- add migration autodetector dependency tracking for custom through models
- resolve string-based through references in schema editor helpers before touching _meta
- add regression coverage for cross-app many-to-many through ordering

## Observed Failure
- Error: `AttributeError: 'str' object has no attribute '_meta'` while running `migrate`.
- Stack trace excerpt:
  ```
  Applying fonte.0001_initial...Traceback (most recent call last):
    ...
    File "/.../django/db/backends/base/schema.py", line 453, in create_model
      if field.remote_field.through._meta.auto_created:
  AttributeError: 'str' object has no attribute '_meta'
  ```

## Reproduction Steps
1. Create apps `fonte`, `variavel`, and `fonte_variavel`.
2. Define `FonteModel` with `variaveis = ManyToManyField("variavel.VariavelModel", through="fonte_variavel.FonteVariavelModel")`.
3. Define `VariavelModel` in `variavel` and `FonteVariavelModel` with foreign keys to both models in `fonte_variavel`.
4. Run `makemigrations` (migration for `fonte` depends only on `variavel`).
5. Run `migrate` and observe the AttributeError above.

Fixes #366

## Testing
- PYTHONPATH=$PWD DJANGO_SETTINGS_MODULE=tests.test_sqlite .venv/bin/python tests/runtests.py migrations.test_autodetector --parallel=1
- .venv/bin/black --check django/db/migrations/autodetector.py django/db/backends/base/schema.py tests/migrations/test_autodetector.py

### CODE REVIEW

**Comment by casey-brooks** [comment]
## Test & Lint Summary
- PYTHONPATH=$PWD DJANGO_SETTINGS_MODULE=tests.test_sqlite .venv/bin/python tests/runtests.py migrations.test_autodetector --parallel=1 ➜ 165 passed, 0 failed, 0 skipped
- .venv/bin/black --check django/db/migrations/autodetector.py django/db/backends/base/schema.py tests/migrations/test_autodetector.py ➜ no formatting changes required

**Comment by rowan-stein** [comment]
Requesting review.

Context:
- Issue: https://github.com/agyn-sandbox/django/issues/366
- Fix ensures migrations autodetector adds dependency on custom ManyToMany `through` model across apps and adds defensive resolution in schema editor for string `through` refs.

Please focus on:
- Correctness of `_get_dependencies_for_foreign_key()` change in `django/db/migrations/autodetector.py`.
- Scope and safety of defensive guards in `django/db/backends/base/schema.py` (create/add/remove/alter paths).
- Regression test coverage in `tests/migrations/test_autodetector.py` validating dependency ordering for cross-app `through`.

CI is not required for this task; local tests have passed.

File: django/db/backends/base/schema.py [hunk 1]
```diff
@@ -450,15 +450,21 @@ def create_model(self, model):
 
         # Make M2M tables
         for field in model._meta.local_many_to_many:
-            if field.remote_field.through._meta.auto_created:
-                self.create_model(field.remote_field.through)
+            through = field.remote_field.through
+            if isinstance(through, str):
+                through = model._meta.apps.get_model(through)
```
Comment (noa-lucent) [inline django/db/backends/base/schema.py]: [major] When  is provided as an app-relative string (e.g.  inside the same app),  raises  because the single-argument form expects a dotted path. These short forms are common in real projects and, if the relation hasn't resolved yet, this guard will now fail with a new exception. Please resolve the string first (e.g. via  to get the  tuple) and feed the two-argument form of . The same adjustment is needed for every new  call in this file.

Comment (noa-lucent) [inline django/db/backends/base/schema.py]: [major] When \\`through\\` is provided as an app-relative string (e.g. \\`through=\"Membership\"\\` inside the same app), `model._meta.apps.get_model(through)` raises `ValueError` because the single-argument form expects a dotted path. These short forms are common in real projects and, if the relation hasn't resolved yet, this guard will now fail with a new exception. Please resolve the string first (for example via `resolve_relation()` to get the `(app_label, model_name)` tuple) and feed the two-argument form of `get_model()`. The same adjustment is needed for every new `get_model()` call in this file.

**Review by noa-lucent — CHANGES_REQUESTED** [review]
Thanks for tackling the cross-app M2M through dependency. The new autodetector dependency logic and regression test both look solid, and the schema guard is heading in the right direction. I left one blocking comment: we need to resolve app-relative through strings (e.g. ) before calling , otherwise those short forms will now explode with a . Please update all the new guard sites to go through  (or equivalent) before calling .

**Comment by casey-brooks** [comment]
Tests & Lint Summary
- PYTHONPATH=tests .venv/bin/python -m tests.runtests --settings=tests.test_sqlite migrations.test_operations --parallel=1 (135 passed, 1 skipped)
- PYTHONPATH=tests .venv/bin/python -m tests.runtests --settings=tests.test_sqlite migrations.test_autodetector --parallel=1 (165 passed, 0 skipped)
- .venv/bin/black django/db/backends/base/schema.py tests/migrations/test_operations.py (no changes)

**Review by noa-lucent — APPROVED** [review]
Thanks for the update—`_resolve_through_model` now resolves both dotted and app-relative strings before hitting `get_model()`, so the original failure mode is gone. The new `test_add_field_m2m_with_relative_through` regression test exercises the schema path for relative through names, and the prior autodetector test continues to cover cross-app dependencies. Everything looks good to me.
